Fix race

hackaugusto opened the pull request:
Fixes the issues with data races. This PR does two main modifications:

1. Move the SQL queries to standalone functions, that operate over the `conn` and its required arguments.
2. Move all the tree states behind a single lock.

Moving the SQL queries to functions is useful because it allows for its reuse. E.g. `apply_block` must validate the new block header data inside a transaction, because the data stored in the DB is not normalized the validation can not be done via SQL constraints, so `apply_block` has to open a transaction, load the last block header from the DB, and perform the validation. Prior to this change the query inside `get_block_header` could only be used inside a new connection, after this change the `sql::get_block_header` can be used from multiple methods.

Moving the tree state behind a single read/write lock allows for atomic updates to the trees. 

This PR is more of a ground work, I will open the `apply_block` as a follow up.

bobbinth reviewed (approved):
Looks good! Thank you! I left just one small nit inline.

One other potential way to group things is to define `StoreApi` as:
```Rust
pub struct StoreApi {
    db: Db,
    state: Arc<State>,
}
```
Where `State` contains the in-memory structures. It looks a bit cleaner to me but also not sure if it will cause any other complications - so, fine to leave as is if you don't think it improves things.

hackaugusto commented:
> Where State contains the in-memory structures. It looks a bit cleaner to me but also not sure if it will cause any other complications - so, fine to leave as is if you don't think it improves things.

I have moved the `db` inside the lock intentionally. The idea is to prevent bugs, with the exception of the `get_block_header` method, all others have to first acquire the lock and then perform queries to the db. As a consequence, that is also the only method that would benefit from having the `db` outside of the lock, and for the time being that doesn't seem to be a method that we are using (I believe we moved all the state needed by the client to sync_state).

Changed it.

hackaugusto commented on store/src/state.rs:
```diff
@@ -0,0 +1,264 @@
+
+const ACCOUNT_DB_DEPTH: u8 = 64;
+
+/// Container for state that needs to be updated atomically.
```
I'm not sure what name you want to use, added a `STRUCTURES` divisor
